src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp

```diff
@@ -26,74 +26,194 @@
 #include <lib/support/CHIPMem.h>
 #include <platform/KeyValueStoreManager.h>
 #include <platform/nxp/k32w/k32w0/K32W0Config.h>
+#include <set>
 #include <string>
 
 #include "PDM.h"
 
+#include <unordered_map>
+
 namespace chip {
 namespace DeviceLayer {
 namespace PersistedStorage {
 
-/* TODO: adjust this value */
-#define MAX_NO_OF_KEYS 255
+/* TODO: adjust these values */
+constexpr size_t kMaxNumberOfKeys  = 20;
+constexpr size_t kMaxKeyValueBytes = 255;
 
 KeyValueStoreManagerImpl KeyValueStoreManagerImpl::sInstance;
 
+/* hashmap having:
+ * 	- the matter key string as key;
+ * 	- internal PDM identifier as value;
+ */
+std::unordered_map<std::string, uint8_t> g_kvs_map;
+
+/* set containing used PDM identifiers */
+std::set<uint8_t> g_key_ids_set;
+
+/* used to check if we need to restore values from flash (e.g.: reset) */
+static bool g_restored_from_flash = false;
+
+CHIP_ERROR RestoreFromFlash()
+{
+    CHIP_ERROR err                        = CHIP_NO_ERROR;
+    uint8_t key_id                        = 0;
+    char key_string_id[kMaxKeyValueBytes] = { 0 };
+    size_t key_string_id_size             = 0;
+    uint8_t pdm_id_kvs                    = chip::DeviceLayer::Internal::K32WConfig::kPDMId_KVS;
+
+    if (g_restored_from_flash)
+    {
+        /* already restored from flash, nothing to do */
+        return err;
+    }
+
+    for (key_id = 0; key_id < kMaxNumberOfKeys; key_id++)
+    {
+        /* key was saved as string in flash (key_string_id) using (key_id + kMaxNumberOfKeys) as PDM key
+         * value corresponding to key_string_id was saved in flash using key_id as PDM key
+         */
+
+        err = chip::DeviceLayer::Internal::K32WConfig::ReadConfigValueStr(
+            chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id + kMaxNumberOfKeys), key_string_id, kMaxKeyValueBytes,
+            key_string_id_size);
+
+        if (err == CHIP_DEVICE_ERROR_CONFIG_NOT_FOUND)
+        {
+            err = CHIP_NO_ERROR;
+            continue;
+        }
+        else if (err != CHIP_NO_ERROR)
+        {
+            break;
+        }
+
+        if (key_string_id_size)
+        {
+            g_key_ids_set.insert(key_id);
+            if (!g_kvs_map.insert(std::make_pair(std::string(key_string_id), key_id)).second)
+            {
+                /* key collision is not expected when restoring from flash */
+                abort();
```

Comment:
Sorry to keep hammering on this, but this is actually the worst thing you can do.  This happens at boot.  If you rely on your runtime to clear flash or shepherd firmware updates (likely), you now have a situation where you will be in a never ending reset loop and can never do those things, and thus bad/inconsistent data in flash will have bricked your device.

What you want to do is log this error for starters.  At a bare minimum, this should be visible in dev builds.  Then you can think about what type of remediation or recovery is appropriate.  Some strategies people employ:

* lock out the broken subsystem
* place the entire device in some global 'degraded' state
* self-recovery / cleanup
* (other things)

Guiding principles:

* don't ignore the error
* consider methods of self-recovery
* but do not prevent manual recovery
* if possible, prevent subsequent (incorrect) execution
* BUT do not self-brick

This can quickly get complicated.  If you just want to log the error for now, that's probably fine.

Reply:
@msandstedt thanks a lot for your review, really appreciated. I replaced the abort with logging for the moment. Depending on how the KVS API evolves, I may add a more complex solution to this.

Reply:
@msandstedt Is this all set in the updated code now?

Reply:
Yes, current code is logging the error, which is the minimum thing we need and can't trigger the boot-time reset loop.